### ISSUE TITLE
Add triu and tril traits to device

### ISSUE THREAD

**Issue by coreylowman** [coreylowman/dfdx#436]
High level interface might look like:

Const shape (triu/tril):
```rust
let val: f32 = ...;
let t: Tensor<Rank2<3, 3>, f32, _> = dev.triu(val);
assert_eq!(t.array(), [
    [val, val, val],
    [0.0, val, val],
    [0.0, 0.0, val],
]);
```

Dynamic shape (triu_like/tril_like):
```rust
let val: f32 = ...;
let t: Tensor<(usize, usize), f32, _> = dev.triu_like(&(3, 3), val);
assert_eq!(t.array(), [
    [val, val, val],
    [0.0, val, val],
    [0.0, 0.0, val],
]);
```

Questions
1. Do we need to support diagonal argument like pytorch?
2. Why did pytorch go with modifying existing tensors instead of creating new ones?

**Comment by jafioti** [comment]
@coreylowman I think the tril / triu in pytorch modifes tensors because it is possible you have a preexisting tensor that you still want to maintain and just take a triangle from, rather than always create a new tensor. So basically for generality purposes. Creating a new tensor and running it through tril is just another line of code, but it's a special case use of tril.

I personally don't think I need the diagonal, but it would be an excellent addition, and likely useful in the future. I'm no cuda expert, but I don't think it would be too much additional complexity to add a diagonal argument in.

**Comment by coreylowman** [comment]
Resolved by #568
